File: hexameter-core/src/main/java/org/codetome/hexameter/core/api/RotationDirection.java
package org.codetome.hexameter.core.api;

/**
 * Represents a right or left angle (60°) of a Hexagon rotation.
 * See: http://www.redblobgames.com/grids/hexagons/#rotation
 */
public enum RotationDirection {

    RIGHT(new RotationCalculator() {
        @Override
        public CubeCoordinate calculate(final CubeCoordinate coord) {
            return CubeCoordinate.fromCoordinates(-coord.getGridZ(), -coord.getGridY());
        }
    }),
    LEFT(new RotationCalculator() {
        @Override
        public CubeCoordinate calculate(final CubeCoordinate coord) {
            return CubeCoordinate.fromCoordinates(-coord.getGridY(), -coord.getGridX());
        }
    });

    private RotationCalculator rotationCalculator;

    RotationDirection(final RotationCalculator rotationCalculator) {
        this.rotationCalculator = rotationCalculator;
    }

    /**
     * Calculates a rotation (right or left) of <code>coord</code>.
     *
     * @param coord coordinate to rotate
     *
     * @return rotated coordinate
     */
    public CubeCoordinate calculateRotation(final CubeCoordinate coord) {
        return rotationCalculator.calculate(coord);
    }

    interface RotationCalculator {
        CubeCoordinate calculate(CubeCoordinate coord);
    }
}

File: mixite2.core/src/commonMain/kotlin/org/hexworks/mixite2/core/api/RotationDirection.kt
package org.hexworks.mixite2.core.api

/**
 * Represents a right or left angle (60°) of a Hexagon rotation.
 * See: http://www.redblobgames.com/grids/hexagons/#rotation
 */
enum class RotationDirection constructor(private val rotationCalculator: RotationCalculator) {

    RIGHT(object : RotationCalculator {
        override fun calculate(coord: org.hexworks.mixite2.core.api.CubeCoordinate): org.hexworks.mixite2.core.api.CubeCoordinate {
            return org.hexworks.mixite2.core.api.CubeCoordinate.fromCoordinates(-coord.gridZ, -coord.gridY)
        }
    }),
    LEFT(object : RotationCalculator {
        override fun calculate(coord: org.hexworks.mixite2.core.api.CubeCoordinate): org.hexworks.mixite2.core.api.CubeCoordinate {
            return org.hexworks.mixite2.core.api.CubeCoordinate.fromCoordinates(-coord.gridY, -coord.gridX)
        }
    });

    /**
     * Calculates a rotation (right or left) of `coord`.
     *
     * @param coord coordinate to rotate
     *
     * @return rotated coordinate
     */
    fun calculateRotation(coord: org.hexworks.mixite2.core.api.CubeCoordinate): org.hexworks.mixite2.core.api.CubeCoordinate {
        return rotationCalculator.calculate(coord)
    }

    internal interface RotationCalculator {
        fun calculate(coord: org.hexworks.mixite2.core.api.CubeCoordinate): org.hexworks.mixite2.core.api.CubeCoordinate
    }
}
